keps/sig-cluster-lifecycle/0029-20180918-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,184 @@
+  with `kubeadm init`.
+- To support extension of the "phases" concept to other kubeadm workflows.
+- To enable re-use of phases across different workflows e.g. the cert phase
+  used by the `kubeadm init` and by the `kubeadm join --master` workflows.
+
+### Non-goals
+
+- This proposal doesn't include any changes of improvements to the actual `kubeadm init`
+  workflow.
+- This proposal doesn't include implementation of workflows different than `kubeadm init`;
+  nevertheless, this proposal introduces a framework that will allow such implementation in future.
+
+## Proposal
+
+### User Stories
+
+- As a kubernetes administrator/IT automation tool, I want to run all the phases of
+  the `kubeadm init` workflow.
+- As a kubernetes administrator/IT automation tool, I want to run only one or some phases
+  of the `kubeadm init` workflow.
+- As a kubernetes administrator/IT automation tool, I want to run all the phases of
+  the `kubeadm init` workflow except some phases.
+
+### Implementation Details
+
+The core of the new phase design consist into a simple, lightweight workflow manager to be used
+for implementing composable kubeadm workflows.
+
+Composable kubeadm workflows are build by an ordered sequence of phases; each phase can have it's
+own, nested, ordered sequence of phases. For instance:
+
+```bash
+  preflight       Run master pre-flight checks
+  certs           Generates all PKI assets necessary to establish the control plane
+    /ca             Generates a self-signed kubernetes CA 
+    /apiserver      Generates an API server serving certificate and key
+    ...
+  kubeconfig      Generates all kubeconfig files necessary to establish the control plane 
+    /admin          Generates a kubeconfig file for the admin to use and for kubeadm itself
+    /kubelet        Generates a kubeconfig file for the kubelet to use.
+    ...
+  ...
+````
+
+Phases implementation will be designed in order to ensure that phases are reusable across different
+workflows thus allowing e.g. reuse of phase `certs` in both `kubeadm init` and
+`kubeadm join --master` workflows.
+
+Each kubeadm workflow will be defined and managed using an `Executor` that basically is a component
+in charge of of running phases according to the given order; nested phases will be executed
+immediately after their parent phase.
+
+The `Executor` will be instantiated by kubeadm commands, and the executor options exposed as
+kubeadm command flags thus allowing to filter the list of phases to be executed. As a consequence
+the `kubeadm init` UX will be modified from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [workflow-flags] [flags]
+```
+
+[workflow-flags] will be a set of optional flags that allows user to select which phase
+of the `kubeadm init` logic to execute (`--only-phases`) or to skip (`--skip-phases`). e.g.
+
+```bash
+kubeadm init --only-phases certs [flags]
```

Comment:
I would really like to see a way to pass in a file containing a list of phases either in addition to replacing the --only-phases flag where you specify phases separated by commas. I think this would provide a nice way to modify and iterate on a list of phases to run without relying on bash history. It would also be nice to say `kubeadm init --phases > phases.txt` then modify the list of phases to be specific.

Reply:
+1 on having this as part of init configuration

Reply:
As agree in kubeadm office hours we are giving the users only the capability to invoke one phase at time